File: src/db/catalog.ts
/**
 * Table capabilities — typed metadata for the DB catalog.
 * Shared by CLI and viewer to determine what operations are allowed per table.
 */
import type { TableName } from "./schema.ts";

// ─── Viewer table names (excludes meta + shadow tables) ─────────────

/** Tables visible in the viewer, ordered by viewer_rank. */
export type ViewerTableName =
	| "canvas_grades"
	| "canvas_assignments"
	| "canvas_submissions"
	| "gh_submissions"
	| "gh_students"
	| "gh_assignments"
	| "students"
	| "assignments";

export interface TableCapability {
	editable: boolean;
	pushable: boolean;
	pullGuarded: boolean;
	viewerRank: number;
	/** If set, only these columns are editable. null = all columns editable. */
	editableColumns: string[] | null;
}

export const TABLE_CAPABILITIES: Record<ViewerTableName, TableCapability> = {
	canvas_grades: {
		editable: true,
		pushable: true,
		pullGuarded: true,
		viewerRank: 0,
		editableColumns: null,
	},
	canvas_assignments: {
		editable: true,
		pushable: true,
		pullGuarded: true,
		viewerRank: 1,
		editableColumns: null,
	},
	canvas_submissions: {
		editable: false,
		pushable: false,
		pullGuarded: false,
		viewerRank: 2,
		editableColumns: null,
	},
	gh_submissions: {
		editable: false,
		pushable: false,
		pullGuarded: false,
		viewerRank: 10,
		editableColumns: null,
	},
	gh_students: {
		editable: true,
		pushable: false,
		pullGuarded: false,
		viewerRank: 11,
		editableColumns: ["excluded"],
	},
	gh_assignments: {
		editable: false,
		pushable: false,
		pullGuarded: false,
		viewerRank: 12,
		editableColumns: null,
	},
	students: {
		editable: false,
		pushable: false,
		pullGuarded: false,
		viewerRank: 20,
		editableColumns: null,
	},
	assignments: {
		editable: false,
		pushable: false,
		pullGuarded: false,
		viewerRank: 21,
		editableColumns: null,
	},
};

// ─── Pushable columns per table ─────────────────────────────────────

export const CANVAS_PUSHABLE = {
	canvas_assignments: ["name", "points_possible", "due_at", "published"] as const,
	canvas_grades: ["posted_grade"] as const,
} as const;

// ─── Derived sets ───────────────────────────────────────────────────

/** Tables hidden from the viewer. */
export const EXCLUDED_TABLES: TableName[] = [
	"meta",
	"canvas_students",
	"_canvas_assignments_synced",
	"_canvas_grades_synced",
];

/** Tables available for export, in order. */
export const EXPORTABLE_TABLES: TableName[] = [
	"students",
	"assignments",
	"gh_students",
	"canvas_students",
	"gh_assignments",
	"canvas_assignments",
	"gh_submissions",
	"canvas_submissions",
	"canvas_grades",
];

// ─── Helper functions ───────────────────────────────────────────────

function getCap(table: string): TableCapability | undefined {
	return TABLE_CAPABILITIES[table as ViewerTableName];
}

export function isEditable(table: string): boolean {
	return getCap(table)?.editable ?? false;
}

export function isPushable(table: string): boolean {
	return getCap(table)?.pushable ?? false;
}

export function isPullGuarded(table: string): boolean {
	return getCap(table)?.pullGuarded ?? false;
}

/** Returns restricted editable columns, or null if all columns are editable. */
export function getEditableColumns(table: string): string[] | null {
	const cap = getCap(table);
	if (!cap?.editable) return null;
	return cap.editableColumns;
}

File: cass/db/catalog.py
"""Shared catalog metadata for user-facing cass data workflows."""

from __future__ import annotations

__docformat__ = "google"

from dataclasses import dataclass


@dataclass(frozen=True)
class TableCapability:
    """Shared table behavior metadata for CLI and viewer workflows."""

    editable: bool = False
    pushable: bool = False
    pull_guarded: bool = False
    viewer_rank: int = 999
    editable_columns: frozenset[str] | None = None


TABLE_CAPABILITIES: dict[str, TableCapability] = {
    "canvas_grades": TableCapability(
        editable=True,
        pushable=True,
        pull_guarded=True,
        viewer_rank=0,
    ),
    "canvas_assignments": TableCapability(
        editable=True,
        pushable=True,
        pull_guarded=True,
        viewer_rank=1,
    ),
    "canvas_submissions": TableCapability(viewer_rank=2),
}

CANVAS_WORKING_TABLES = {
    name for name, capability in TABLE_CAPABILITIES.items() if capability.editable
}
PULL_GUARDED_TABLES = {
    name for name, capability in TABLE_CAPABILITIES.items() if capability.pull_guarded
}


def get_table_capability(table: str) -> TableCapability:
    """Return shared behavior metadata for a table."""
    return TABLE_CAPABILITIES.get(table, TableCapability())


CANVAS_PUSHABLE: dict[str, set[str]] = {
    "canvas_assignments": {"name", "points_possible", "due_at", "published"},
    "canvas_grades": {"posted_grade"},
}

EXPORTABLE_TABLES = (
    "canvas_students",
    "canvas_assignments",
    "canvas_submissions",
    "canvas_grades",
)

IMPORT_SIGNATURES: dict[str, set[str]] = {
    "canvas_grades": {"canvas_user_id", "canvas_assignment_id"},
    "canvas_submissions": {
        "canvas_user_id",
        "canvas_assignment_id",
        "submitted",
    },
}
